persistent/persistence.py

```diff
@@ -331,36 +361,81 @@ def __reduce__(self):
     def _p_activate(self):
         """ See IPersistent.
         """
-        before = self.__flags
-        if self.__flags is None or self._p_state < 0: # Only do this if we're a ghost
-            self.__flags = 0
-            if self.__jar is not None and self.__oid is not None:
-                try:
-                    self.__jar.setstate(self)
-                except:
-                    self.__flags = before
-                    raise
+        oga = _OGA
+        before = oga(self, '_Persistent__flags')
+        if before is None: # Only do this if we're a ghost
+            # Begin by marking up-to-date in case we bail early
+            _OSA(self, '_Persistent__flags', 0)
+            jar = oga(self, '_Persistent__jar')
+            if jar is None:
+                return
+            oid = oga(self, '_Persistent__oid')
+            if oid is None:
+                return
+
+            # If we're actually going to execute a set-state,
+            # mark as changed to prevent any recursive call
+            # (actually, our earlier check that we're a ghost should
+            # prevent this, but the C implementation sets it to changed
+            # while calling jar.setstate, and this is observable to clients).
+            # The main point of this is to prevent changes made during
+            # setstate from registering the object with the jar.
+            _OSA(self, '_Persistent__flags', CHANGED)
+            try:
+                jar.setstate(self)
+            except:
+                _OSA(self, '_Persistent__flags', before)
+                raise
+            else:
+                # If we succeed, no matter what the implementation
+                # of setstate did, mark ourself as up-to-date. The
+                # C implementation unconditionally does this.
+                _OSA(self, '_Persistent__flags', 0) # up-to-date
+
+    # In the C implementation, _p_invalidate winds up calling
+    # _p_deactivate. There are ZODB tests that depend on this;
+    # it's not documented but there may be code in the wild
+    # that does as well
```

Comment:
I'm not sure that it's a matter of fixing them. The test is very particular about the order in which things are called, it doesn't seem to be accidental. That leads me to believe that even if it isn't documented it's behaviour that people could have been relying on.